include/rdma/fi_profile.h

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2023 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef FI_PROFILE_H
+#define FI_PROFILE_H
+
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+
+#include <rdma/fi_domain.h>
+
+/*
+ * pre-defined profiling variables
+ */
+enum {
+	FI_VAR_UNEXP_MSG_CNT,    // type: uint64_t
+};
+
+/*
+ * pre-defined events which can change the profiling variables
+ */
+enum {
+	FI_EVENT_UNEXP_MSG_RECVD,    // var = FI_VAR_UNEXP_MSG_CNT
+	FI_EVENT_UNEXP_MSG_MATCHED,  // var = FI_VAR_UNEXP_MSG_CNT
+};
+
+struct fi_profile_var_desc {
+	uint32_t id;
+	enum fi_datatype type;
+	size_t size;
+	const char *name;
+	const char *desc;
+};
+
+struct fi_profile_event_desc {
+	uint32_t id;
+	const char *name;
+	const char *desc;
+};
```

Comment:
Let's merge the above structs into just fi_profile_desc.  As mentioned separately, this gives us some versioning protection on the data that's passed into a callback.  We can add new values to fi_datatype to define specific structures for libfabric defined data as needed, as well as something like FI_PROV_TYPE to indicate that it's a provider specific structure.  New types would carve out a portion of the enum space, similar to what's been done for collectives.

This also makes it easier to have a value be both a variable and an event, such that the app could monitor changes to the variable.